### DISCUSSION TITLE
pyinstaller refuses to build a standalone executable even though i used the --onefile option

### DISCUSSION THREAD

**Discussion by Fares-Rbd** [pyinstaller/pyinstaller#7788] [Help]
i have a python project consisting of multiple python files and that uses external libraries like PyQt5 (the user interface is made with Qt Designer) the project works just fine when i run it as a python file, however the problems begin when i try to convert it into an .exe file, i need this project to be converted into a single .exe file but whenever i try doing that some problem occurs.
  What i tried so far :
- used `pyinstaller --onefile main.py`, this resulted in the following interface that is basically the one fresh out of Qt designer and before i apply the logic and other modifications in my python code, here is an image:
![image](https://github.com/pyinstaller/pyinstaller/assets/140640835/1686736a-f352-4845-a1a2-bd8648d123a7)

- Next i tried using a .spec file which actually generated a working project but the problem is that in the dist folder instead of a standalone executable i found a folder full of files among them is an executable that works but as i mentioned i need this to be a single executable file, here is the .spec file i used : [mainspec.txt](https://github.com/pyinstaller/pyinstaller/files/12175145/mainspec.txt)
- then i asked on StackOverFlow and a user told me that if i want to generate a standalone executable i should get rid of the COLLECT() statement and i did but that resulted in some warnings , an empty dist folder and a build folder with a bunch of files among them is the executable that when clicked on pops this error then crashes: 
![image](https://github.com/pyinstaller/pyinstaller/assets/140640835/e4a88589-dd6e-4627-bbde-28dd07e4d24e)

here is the warning log file : [warn-main.txt](https://github.com/pyinstaller/pyinstaller/files/12175239/warn-main.txt)

**Comment by rokm** [comment]
When you ran `pyinstaller --onefile main.py`, PyInstaller generated a spec file for `onefile` build called `main.spec`. Why aren't you just editing that one?

The attached `mainspec.txt` is for `onedir` build, so of course it a) generates a onedir frozen application and b) stops working if you remove `COLLECT` from it.

There are subtle differences between `onefile` and `onedir` build .spec files, so unless you know what you are doing, I advise you against trying to convert one into the other, but instead create a new spec file for the desired build type, and edit that.


PS: your `mainspec.txt` seems to specify multiple .py files as entry point scripts, which is incorrect; only the actual entry-point script (main.py) should be specified there:
``` 
a = Analysis(
    ['main.py', 'ui_interface.py', 'utilities.py', 'data.py'],
    pathex=[],
```

**Reply by Fares-Rbd** [reply]
Yes i am working with the `main.spec` file generated when i used `pyinstaller --onefile main.py` i just sent it as a txt file called `mainspec.txt` because github doesn't allow .spec files, as for the multiple entry points i think chatgbt added them because my main.py file uses data and functions from those other python files , i generated the following new main.spec file after running `pyinstaller --onefile main.py` : [mainspec.txt](https://github.com/pyinstaller/pyinstaller/files/12176629/mainspec.txt) what should i change in this main.spec file so that the program works ? 

PS: the multiple entry points that you specified didnt make a problem, using that first main.spec file i sent i was able to generate a working program but as i mentionned, the problem there is that it generates a whole folder instead of a standalone executable.

**Reply by rokm** [reply]
Take a look at the first `mainspec.txt` that you attached, and copy the relevant entries from it - the `datas=...` and `hiddenimports=...` arguments to `Analysis`, and probably `name=...` and `icon=...` arguments to `EXE`.

**Reply by Fares-Rbd** [reply]
I did as you told me and wrote this new main.spec file : [mainspec.txt](https://github.com/pyinstaller/pyinstaller/files/12178727/mainspec_3.txt) yet the problem still persists, it generates a standalone executable but it does not work, the interface is the same as the one show in the picture i sent above, here is the warning log for the program generated with that main.spec file maybe the information in there will be useful [warn-main.txt](https://github.com/pyinstaller/pyinstaller/files/12178734/warn-main.txt)

Appreciate your help.

**Reply by rokm** [reply]
I don't think I undestand what the problem is, then - what is wrong with the UI in the screenshot?

**Reply by Fares-Rbd** [reply]
okay i will explain again, begining with how the interface should look like: 
[demo.webm](https://github.com/pyinstaller/pyinstaller/assets/140640835/24a36b5c-e77d-4221-a45c-2e81e675cc78)
this is how the program should look like once i build, and if i build it without the onefile option it works just fine and as intended, the problem begins when i use the onefile option (which i must use) every time i build it using the one file option i get something like this even if i add all the data and hidden imports : 
[demo2.webm](https://github.com/pyinstaller/pyinstaller/assets/140640835/6978ea37-1ab9-4899-a6d4-5694e1b28c8c)
So the program generated in the 2nd video does not work, the buttons dont work the interface looks like the one straight out of my Qt designer file as if all the logic, animation and functionality i added in my python code where not included in the standalone executable that was built which is weird because as i said, if i build a whole folder instead of a standalone executable the program works just fine and somehow not the case when i use the onefile option ? 

Edit: the program that was shown in `demo.webm` (the one that works as intended) and the program shown `demo2.webm` (the one that does not work) where built using almost the same `main.spec` file, the only difference is that in the working program i added the COLLECT(...) statement which forces it to generate a whole folder but gives me a working program, once i removed the COLLECT(...) statement it generates a standalone executable that does not work

**Reply by rokm** [reply]
Try building with console enabled (`console=True` to `EXE`), and see if you get any error messages printed on console. 

If it works in onedir but not in onefile, your code might be assuming that some resources are available in the current working directory, and that assumption is broken when you run in onefile mode (because resources are not next to executable, but extracted in temporary directory).

**Reply by Fares-Rbd** [reply]
I just tried that and here is the console's output : [console.txt](https://github.com/pyinstaller/pyinstaller/files/12185619/console.txt)
I also just remembered that i used a custom library to enable the animations that are shown in the `demo.webm` video, the library is called `Custom_Widgets` which i think is based on `PyQt5` if i remember correctly, and to enable the animations i had to write some json code in a `style.json` file which is mentionned in the datas parameter in the `main.spec` file, maybe the `Custom_Widgets` library is the cause of that problem ? 

PS: i added the `Custom_Widgets` library in the hidden imports section and rebuilt the program, still the same result and here is the console output when i run that build : [console2.txt](https://github.com/pyinstaller/pyinstaller/files/12185740/console.txt)
